File: trie-service/trie-service.py
"""
    Definition of TrieNode:
    class TrieNode:
    def __init__(self):
    # <key, value>: <Character, TrieNode>
    self.children = collections.OrderedDict()
    self.top10 = []
    """


class TrieService:
    
    def __init__(self):
        self.root = TrieNode()
    
    def get_root(self):
        # Return root of trie root, and
        # lintcode will print the tree struct.
        return self.root
    
    # @param {str} word a string
    # @param {int} frequency an integer
    # @return nothing
    def insert(self, word, frequency):
        # Write your your code here
        p = self.root
        for c in word:
            if c not in p.children:
                p.children[c] = TrieNode()
            p = p.children[c]
            if len(p.top10) == 10:
                p.top10[9] = max(p.top10[9], frequency)
            else:
                p.top10.append(frequency)
            p.top10.sort(reverse=True)

File: trie-service/trie-service.cc
/**
 * Definition of TrieNode:
 * class TrieNode {
 * public:
 *     TrieNode() {}
 *     map<char, TrieNode*> children;
 *     vector<int> top10;
 * };
 */
class TrieService {
    private:
    TrieNode* root;
    
    public:
    TrieService() {
        root = new TrieNode();
    }
    
    TrieNode* getRoot() {
        // Return root of trie root, and
        // lintcode will print the tree struct.
        return root;
    }
    
    void insert(string& word, int frequency) {
        // Write your code here
        TrieNode *p = root;
        for (int i = 0; word[i]; ++i) {
            if (p->children.find(word[i]) == p->children.end()) {
                p->children[word[i]] = new TrieNode();
            }
            vector<int> &top = p->children[word[i]]->top10;
            if (top.size() == 10) {
                if (top[9] < frequency) {
                    top[9] = frequency;
                }
            } else {
                top.push_back(frequency);
            }
            sort(top.begin(), top.end());
            reverse(top.begin(), top.end());
            p = p->children[word[i]];
        }
    }
};
